gensim/models/nmf.py

```diff
@@ -487,53 +562,50 @@ def update(self, corpus, chunks_as_numpy=False):
         chunk_idx = 1
 
         for _ in range(self.passes):
-            for chunk in utils.grouper(
-                corpus, self.chunksize, as_numpy=chunks_as_numpy
-            ):
+            for chunk in utils.grouper(corpus, self.chunksize, as_numpy=False):
                 self.random_state.shuffle(chunk)
-                v = matutils.corpus2csc(chunk, len(self.id2word)).tocsr()
-                self._h, self._r = self._solveproj(
-                    v, self._W, r=self._r, h=self._h, v_max=self.v_max
-                )
-                h, r = self._h, self._r
+
+                if isinstance(corpus, scipy.sparse.csc.csc_matrix):
```

Comment:
Hmm, how does this work in combination with `grouper` above? The grouper groups per rows, but CSC doesn't allow row-wise slicing?

Reply:
Yes, it's a bug, I'll fix it

Reply:
![image](https://user-images.githubusercontent.com/3992246/52011077-167fd600-24e8-11e9-982e-a2ad6f90c427.png)

Fixed, faces look much better now. :)

Reply:
Hmm, so if I understand correctly, previously we were interpreting documents as features and vice versa, while shuffling the features?

How did that even work? Something must be wrong with the eval if that worked (enough) to pass all human & automated checks.

Reply:
Yes, but only in the case of csc in Olivietti faces dataset, other examples use iterable of word counts as input.

I think what happen is one bug interfered with another an it worked. :neutral_face: 

Specifically, in the notebook I transposed input faces matrix twice, thus passing dataset with shape `(n_faces, n_pixels)` to the model, where it was split to the lists of rows of shape `(chunksize, n_pixels)`, and [here](https://github.com/RaRe-Technologies/gensim/blob/7458fa566aea57edcdc249b7feb93578867b4c6c/gensim/models/nmf.py#L570) I took the first row in a batch and transposed it, thus getting `v` with shape `(n_pixels, 1)`, which is valid dimensions.

The effect of this was that NMF were training only on the first face of every batch, that's why the quality of faces were worse (only 1/chunksize of examples were actually used).

Reply:
Beautiful! Such tricky double-bugs can be hard to find :)

Can we do anything to prevent similar bugs in the future? More unit tests / asserts?

Reply:
Sure! We'll just need to check shapes of internal matrices after an update.